servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/MulticastSingle.java

```diff
@@ -0,0 +1,371 @@
+                    }
+                    break;
+                }
+            }
+        }
+
+        @Override
+        public void onSubscribe(final Cancellable cancellable) {
+            delayedCancellable.delayedCancellable(cancellable);
+        }
+
+        @Override
+        public void onSuccess(@Nullable final T result) {
+            safeTerminalStateReset(result, null);
+            terminate(new TerminalSubscriber<>(null, result), sub -> sub.onSuccess(result));
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            safeTerminalStateReset(null, t);
+            terminate(new TerminalSubscriber<>(t, null), sub -> sub.onError(t));
+        }
+
+        private void terminate(TerminalSubscriber<T> terminalSubscriber, Consumer<Subscriber<? super T>> terminator) {
+            @SuppressWarnings("unchecked")
+            final Subscriber<? super T>[] newSubs = (Subscriber<? super T>[]) Array.newInstance(Subscriber.class, 1);
+            newSubs[0] = terminalSubscriber;
+            for (;;) {
+                final Subscriber<? super T>[] subs = subscribers;
+                if (newSubscribersUpdater.compareAndSet(this, subs, newSubs)) {
+                    Throwable delayedCause = null;
+                    for (final Subscriber<? super T> subscriber : subs) {
+                        try {
+                            terminator.accept(subscriber);
+                        } catch (Throwable cause) {
+                            delayedCause = catchUnexpected(delayedCause, cause);
+                        }
+                    }
+                    if (delayedCause != null) {
+                        throwException(delayedCause);
+                    }
+                    break;
+                }
+            }
+        }
+
+        private void safeTerminalStateReset(@Nullable T value, @Nullable Throwable t) {
+            Completable completable;
+            try {
+                completable = terminalResubscribe.apply(value, t);
+            } catch (Throwable cause) {
+                LOGGER.warn("terminalStateReset {} threw", terminalResubscribe, cause);
+                completable = Completable.never();
+            }
+            completable.afterFinally(() -> state = new State()).subscribe();
+        }
+    }
+
+    private static final class ConcurrentOnSubscribeSubscriber<T> implements Subscriber<T> {
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<ConcurrentOnSubscribeSubscriber, Object> stateUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(ConcurrentOnSubscribeSubscriber.class, Object.class, "state");
+        private static final Object INIT = new Object();
+        private static final Object INVOKING_ON_SUBSCRIBE = new Object();
+        private static final Object WAITING_FOR_TERMINAL = new Object();
+        private static final Object TERMINATED = new Object();
```

Comment:
Consider using enum for these states, it's easier to read them in the heap dump then